### DISCUSSION TITLE
reverse proxy - serve bb from a specific path

### DISCUSSION THREAD

**Discussion by dncpax** [Budibase/budibase#3023] [Help]
Hi there (again).

I don't mean to carpet question the bb team... but as I progress I do find difficulties and have no other option than to pester you good folks...

Sooo... I'm trying to do what the reverse proxy doc page says:
https://docs.budibase.com/self-hosting/hosting-settings/advanced-options/reverse-proxy

> Where location / has been set you can change the slash to a path name if you desire to have your Budibase app server running on a specific path rather than the whole domain or a sub-domain.

But I'm not seeing how this can work, since bb uses absolute paths to load assets like "/builder/assets/index.32c8bc7d.js".

Even if I setup my reverse proxy to use a path (eg server/budibase/builder), bb will return absolute url's that don't include this path.

At first I thought the Platform URL was for this purpose, but it seems it's not...

May be there's a config for envoy that I'm not aware of?

What to do???

Thanks.

**Comment by dncpax** [comment]
I used oubound rules to add a subpath on the reverse proxy. I have done this before. It's a pain but works.

The problem is budibase complains of unknown route:
```
Uncaught Error: Route could not be found for "/subpath/builder".
    at Oe (vendor.60887453.js:1)
    at Object.r [as updatePage] (vendor.60887453.js:1)
    at vendor.60887453.js:1

```

I'm starting to think this is not possible... which is a deal breaker for my use case. I just cannot afford creating a sub-domain for every web app...

Any help would be really appreciated.

I also played around with envoy.yaml which would allow to use a simpler reverse proxy config, but the same error occurs.

Thanks.

**Comment by shogunpurple** [comment]
Hi @dncpax - thank you for the detailed write up on this. I believe you are correct - before, we used to serve static assets on a specific path, that the node server served directly. We then moved to envoy and minIO, which changed things.

The error that you are seeing there appears to be a front end issue with our routing library - the front end is trying to serve the route at `/subpath/builder`. You would need to have your reverse proxy strip the `/subpath` before the request gets through. 

What are your exact requirements? Are you trying to serve each individual app on a separate path, so that users can go directly to them and bypass the portal?

**Reply by dncpax** [reply]
Hi @shogunpurple. The requirements are very simple really - I concentrate all apps in a single http server. This acts as a reverse proxy for each app. For instance I have jenkins set up like this, redash, etc - http_server/jenkins, http_server/app2, http_server/app3, etc.
So budibase would also be accessed like this - on a subpath on this http server, and reverse proxyied to budibase docker_ip:10000.
Instead of telling my users that budibase url is something like internal_ip:10000, I tell them it's the usual http_server/budibase.

I don't really see how I could "strip the /subpath before the request gets through"?

I mean, at this time I have configured the reverse proxy to replace all absolute paths in index.html like /assets or /builder to /budibase/assets or budibase/builder, so the browser can request files correctly to http_server/budibase. This then proxyies the requests to docker_ip:10000. I don't see how I can strip anything really.
This is what substring module in nginx usually is used for, when apps don't have subpath configuration support:
browser --> http_server/budibase/index.html --> docker_ip:10000/index.html --> http_server replaces all absolute links with /budibase/* -->browser
browser --> http_server/budibase/builder/assets/index.32c8bc7d.js --> docker_ip:10000/builder/assets/index.32c8bc7d.js -->http_server -->browser

So the index.32c8bc7d.js is requested from budibase with the original url. I think the problem is the routing lib on the browser has a list of known routes that don't include the subpath, hence the error when the browser requests something in the subpath.

I hope I managed to be clear enough...

Thanks.

**Reply by shogunpurple** [reply]
Hi @dncpax - that makes sense. So when I referred to rewriting the paths.

```
rewrite /subpath(.*) $1 break;
```

Which would strip the prefix before it forwards the request to the target backend. If you add this rewrite, it should remove the `subpath` prefix, before it reaches the server.

Would this help for your use case?

**Reply by dncpax** [reply]
I don't think it helps, that's what i'm already doing. I think the error occurs because the browser is requesting url's that don't match the predefined routes. It happens before ever reaching the server. This is my understanding but I may be wrong...

At the moment I don't send any requests to bb with subpath.
I insert the subpath while intercepting the responses from budibase to the browser.

**Reply by dncpax** [reply]
Let me go and double-check this...

**Reply by dncpax** [reply]
ok, double-checked. As you can see the url's reaching bb are without subpaths:
`OldUrl="/budibase/builder/assets/vendor.d75d5eb7.css", NewUrl="http://192.168.2.25:10000/builder/assets/vendor.d75d5eb7.css"`

192.168.2.25 is the docker_ip...

The error that shows in the browser console:
```
Uncaught Error: Route could not be found for "/budibase/builder".
    at Oe (vendor.60887453.js:1)
    at Object.r [as updatePage] (vendor.60887453.js:1)
    at vendor.60887453.js:1
```

Following the error, we get to this code section:
```
function Oe(t, e=!1) {
    t = me.urlTransform.remove(t);
    let {pathname: a, search: r} = xe(t).url;
    const n = y(ke)
      , i = n.find((t=>a === t.meta.name)) || n.find((t=>a.match(t.regex)));
    if (!i)
        throw new Error(`Route could not be found for "${a}".`);  **<---- this is the error message we see**
    const o = e ? Object.create(i) : i
      , {route: l, redirectPath: c, rewritePath: s} = Te(o, n);
    return s && (({pathname: a, search: r} = xe(we(s, l.params)).url),
    c && (l.redirectTo = we(c, l.params || {}))),

```

I am not familiar with routify which I'm guessing bb is using, but it seems routes are defined at build time... not at runtime...

**Reply by dncpax** [reply]
just pinging @shogunpurple ...

**Reply by shogunpurple** [reply]
Hi @dncpax,

Hm so yes, it seems like routify renders that route with the sub path included.

I will need to try this myself with an NGINX config tomorrow and see if I can get it to work - I will update this discussion as I do. 

Thank you!

**Reply by dncpax** [reply]
Hi @shogunpurple. Ok, thanks for testing this.

**Reply by shogunpurple** [reply]
Hi @dncpax, thanks a lot for your patience on this one. So I've had some time to play around with this, and you are right, it's not possible to render a route for each app while you also use a subpath for `/budibase` without some very esoteric NGINX config. By far the best way to achieve this, allowing you to have certain apps on subpaths and use budibase self hosted with a simpler NGINX config is to do the following. This is how I run my instance, too. You would only have to add one subdomain for the builder, as opposed to one for every app.

- Add a subdomain to point to your budibase server, for example `budibase.yourserver.com`. This serves as the root, and alleviates all the problems you were having with fetching
- Add path rules for each app, and have NGINX point to the app ID directly

Here's what this might look like in NGINX, if you wanted to show a specific app at `https://budibase.yourserver.com/myapp`

```
server {
    listen 443 ssl;
    listen [::]:443 ssl;

    server_name budibase.yourserver.*;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    location / {
        include /config/nginx/proxy.conf;
        set $upstream_app budibase_proxy-service_1;
        set $upstream_port 10000;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;

    }

    location /myapp {
        include /config/nginx/proxy.conf;
        set $upstream_app budibase_proxy-service_1;
        set $upstream_port 10000;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port/app_dev_cec9a37156774e43a9ad343aac18153f;
    }

    # then add more apps here as you need.
}
```

I know this isn't the subpath method, `/budibase`, but I would recommend that you run budibase with this configuration. Budibase relies heavily on relative paths that go directly back to the server, so this removes all of those concerns by just using a single subdomain. Hope this helps!

**Reply by faizalheesyam** [reply]
> * Add a subdomain to point to your budibase server, for example `budibase.yourserver.com`. This serves as the root, and alleviates all the problems you were having with fetching
> * Add path rules for each app, and have NGINX point to the app ID directly

here i am, facing the same problem few months after this discussion..

is there any "nginx" way to solve the relative path issues when doing this? other suggestion besides using subdomains?

my situation, our dev server still using ip address.. and all these time, we simply proxy our stuffs like:
 `http://XX.XX.XX.XX/metabase` -->  `http://XX.XX.XX.XX:3000`
 `http://XX.XX.XX.XX/api` -->  `http://XX.XX.XX.XX:5001`
...etc

now i really hope budibase could simply work with:
 `http://XX.XX.XX.XX/budibase` -->  `http://XX.XX.XX.XX:10000`

the reverse proxying has no problem, i think its the budibase itself need some changes with the way their routing was build.. when i visit `http://XX.XX.XX.XX/budibase`, it redirects to  `http://XX.XX.XX.XX/builder` which gave me a 404 page.. 

tried to manually add 'budibase' in the path like `http://XX.XX.XX.XX/budibase/builder`, but nginx gave me *no healthy upstream* instead..

any nginx expert can guide me with this?

**Reply by dbudano** [reply]
I have EXACTLY the same problem as @faizalheesyam except that I'm using Traefik as a reverse proxy.

Any update one year later ?

**Reply by jheaff1** [reply]
I'm having the same issues. I'd like to simply host budibase at http://xxxxx/budibase using nginx as a reverse proxy, yet budibase redirects to http://xxxxx/builder rather than say http://xxxx/budibase/builder

**Reply by JayP718** [reply]
@jheaff1 Have you had any luck with this?

**Reply by jheaff1** [reply]
@JayP718 Sadly not

**Reply by zzdpavo** [reply]
same problem here...sadly. any update @shogunpurple @dncpax ?

**Comment by dncpax** [comment]
Hi @shogunpurple. Thanks for looking into this. 
I'll have a go as soon as I can and report back!
